pymc_marketing/mmm/base.py

```diff
@@ -472,19 +481,491 @@ def compute_channel_contribution_original_scale(self) -> DataArray:
+        )
+
+        ax.text(
+            x_mean,
+            ax.get_ylim()[1] / 1.25,
+            f"α: {alpha_limit:.5f}",
+            fontsize=9,
+            bbox=dict(facecolor="white", edgecolor="black", boxstyle="round,pad=0.5"),
+        )
+
+        ax.set(xlabel="Spent", ylabel="Contribution")
+        ax.legend()
+
+    def optimize_channel_budget_for_maximum_contribution(
+        self,
+        method: str,
+        total_budget: int,
+        budget_bounds: Optional[Dict[str, Tuple[float, float]]] = None,
+        *,
+        parameters: Dict[str, Tuple[float, float]],
+    ) -> pd.DataFrame:
+        """
+        Optimize the allocation of a given total budget across multiple channels to maximize the expected contribution.
+
+        The optimization is based on the method provided, where each channel's contribution
+        follows a saturating function of its allocated budget. The function seeks the budget allocation
+        that maximizes the total expected contribution across all channels. The method can be either 'sigmoid' or 'michaelis-menten'.
+
+        Parameters
+        ----------
+        total_budget : int, required
+            The total budget to be distributed across channels.
+        method : str, required
+            The method used to fit the contribution & spent non-linear relationship. It can be either 'sigmoid' or 'michaelis-menten'.
+        parameters : Dict, required
+            A dictionary where keys are channel names and values are tuples (L, k) representing the
+            parameters for each channel based on the method used.
+        budget_bounds : Dict, optional
+            An optional dictionary defining the minimum and maximum budget for each channel.
+            If not provided, the budget for each channel is constrained between 0 and its L value.
+
+        Returns
+        -------
+        DataFrame
+            A pandas DataFrame containing the allocated budget and contribution information.
+
+        Raises
+        ------
+        ValueError
+            If any of the required parameters are not provided or have an incorrect type.
+        """
+        if not isinstance(budget_bounds, (dict, type(None))):
+            raise TypeError("`budget_ranges` should be a dictionary or None.")
+
+        if not isinstance(total_budget, (int, float)):
+            raise ValueError(
+                "The 'total_budget' parameter must be an integer or float."
+            )
+
+        if not parameters:
+            raise ValueError(
+                "The 'parameters' argument (keyword-only) must be provided and non-empty."
+            )
+
+        return budget_allocator(
+            method=method,
+            total_budget=total_budget,
+            channels=list(self.channel_columns),
+            parameters=parameters,
+            budget_ranges=budget_bounds,
+        )
+
+    def compute_channel_curve_parameters_original_scale(self, method: str) -> Dict:
```

Comment:
I will rename this to `compute_channel_curve_optimization_parameters_original_scale` so that people do not think these are model parameters (adstock or saturation)